libtenzir/src/tql2/ast.cpp

```diff
@@ -169,6 +170,51 @@ auto to_operand(const ast::expression& x) -> std::optional<operand> {
     });
 }
 
+// @brief Optimize x > now() +- $y <=> x > $now +- $y and x > now() +- $y
+auto optimize_now(const ast::expression& left,
+                  const relational_operator& rel_op,
+                  const ast::expression& right) -> std::optional<expression> {
+  constexpr auto f = [](const ast::expression& l, const ast::binary_op& op,
+                        const ast::expression& r) -> std::optional<operand> {
+    if (op != ast::binary_op::add and op != ast::binary_op::sub) {
+      return std::nullopt;
+    }
+    // TODO: Evaluate unary_expr to a constant duration
+    auto expr = std::get_if<ast::constant>(r.kind.get());
+    auto call = std::get_if<ast::function_call>(l.kind.get());
+    if (not(expr and call and call->fn.path[0].name == "now")) {
+      return std::nullopt;
+    }
+    auto y = std::get_if<duration>(&expr->value);
+    if (not y) {
+      return std::nullopt;
+    }
+    if (op == ast::binary_op::add) {
+      return operand{data{time::clock::now() + *y}};
+    }
+    TENZIR_ASSERT(op == ast::binary_op::sub);
+    return operand{data{time::clock::now() - *y}};
+  };
+  auto field = to_field_extractor(left);
+  const auto* import_time = std::get_if<ast::meta>(left.kind.get());
+  if (not(field
+          or (import_time and import_time->kind == ast::meta::import_time))) {
+    return std::nullopt;
+  }
+  if (auto expr = std::get_if<ast::binary_expr>(right.kind.get())) {
+    if (auto op = f(expr->left, expr->op.inner, expr->right);
+        op or (op = f(expr->right, expr->op.inner, expr->left))) {
```

Comment:
We are again swapping left and right, but this changes the meaning of the binary expression.